File: backend/src/main/java/com/pgvaale/backend/controller/TiffinController.java
package com.pgvaale.backend.controller;

import com.pgvaale.backend.dto.MenuDTO;
import com.pgvaale.backend.dto.TiffinDashboardDTO;
import com.pgvaale.backend.dto.UserTiffinDTO;
import com.pgvaale.backend.entity.Tiffin;
import com.pgvaale.backend.entity.UserTiffin;
import com.pgvaale.backend.repository.TiffinRepository;
import com.pgvaale.backend.service.TiffinService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;
import java.util.Optional;

@CrossOrigin(origins = "http://localhost:3000")
@RestController
@RequestMapping("/api/tiffin")
public class TiffinController {
    
    @Autowired
    private TiffinService tiffinService;
    
    @Autowired
    private TiffinRepository tiffinRepository;
    
    // Dashboard
    @GetMapping("/dashboard")
    public ResponseEntity<?> getDashboard() {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            
            // Get tiffin ID from username (you might need to adjust this based on your auth setup)
            Long tiffinId = getTiffinIdFromUsername(username);
            TiffinDashboardDTO dashboard = tiffinService.getTiffinDashboard(tiffinId);
            return ResponseEntity.ok(dashboard);
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error fetching dashboard: " + e.getMessage());
        }
    }
    
    // Menu Management
    @PostMapping("/menu")
    public ResponseEntity<?> createMenu(@RequestBody MenuDTO menuDTO) {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            Long tiffinId = getTiffinIdFromUsername(username);
            menuDTO.setTiffinId(tiffinId);
            
            MenuDTO createdMenu = tiffinService.createMenu(menuDTO);
            return ResponseEntity.ok(createdMenu);
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error creating menu: " + e.getMessage());
        }
    }
    
    @PutMapping("/menu/{menuId}")
    public ResponseEntity<?> updateMenu(@PathVariable Long menuId, @RequestBody MenuDTO menuDTO) {
        try {
            MenuDTO updatedMenu = tiffinService.updateMenu(menuId, menuDTO);
            return ResponseEntity.ok(updatedMenu);
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error updating menu: " + e.getMessage());
        }
    }
    
    @DeleteMapping("/menu/{menuId}")
    public ResponseEntity<?> deleteMenu(@PathVariable Long menuId) {
        try {
            tiffinService.deleteMenu(menuId);
            return ResponseEntity.ok("Menu deleted successfully");
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error deleting menu: " + e.getMessage());
        }
    }
    
    @GetMapping("/menu")
    public ResponseEntity<?> getWeeklyMenu() {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            Long tiffinId = getTiffinIdFromUsername(username);
            
            List<MenuDTO> menus = tiffinService.getWeeklyMenu(tiffinId);
            return ResponseEntity.ok(menus);
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error fetching menu: " + e.getMessage());
        }
    }
    
    @GetMapping("/menu/{dayOfWeek}")
    public ResponseEntity<?> getMenuByDay(@PathVariable String dayOfWeek) {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            Long tiffinId = getTiffinIdFromUsername(username);
            
            MenuDTO menu = tiffinService.getMenuByDay(tiffinId, dayOfWeek);
            if (menu != null) {
                return ResponseEntity.ok(menu);
            } else {
                return ResponseEntity.notFound().build();
            }
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error fetching menu: " + e.getMessage());
        }
    }
    
    // Request Management
    @GetMapping("/requests")
    public ResponseEntity<?> getRequests(@RequestParam(required = false) String status) {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            Long tiffinId = getTiffinIdFromUsername(username);
            
            UserTiffin.RequestStatus requestStatus = null;
            if (status != null && !status.isEmpty()) {
                requestStatus = UserTiffin.RequestStatus.valueOf(status.toUpperCase());
            }
            
            List<UserTiffinDTO> requests = tiffinService.getTiffinRequests(tiffinId, requestStatus);
            return ResponseEntity.ok(requests);
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error fetching requests: " + e.getMessage());
        }
    }
    
    @PostMapping("/requests/{requestId}/status")
    public ResponseEntity<?> updateRequestStatus(@PathVariable Long requestId, @RequestBody Map<String, String> request) {
        try {
            String status = request.get("status");
            UserTiffin.RequestStatus requestStatus = UserTiffin.RequestStatus.valueOf(status.toUpperCase());
            
            UserTiffinDTO updatedRequest = tiffinService.updateRequestStatus(requestId, requestStatus);
            return ResponseEntity.ok(updatedRequest);
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error updating request status: " + e.getMessage());
        }
    }
    
    // Profile
    @GetMapping("/profile")
    public ResponseEntity<?> getProfile() {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            
            // You'll need to implement this based on your auth setup
            return ResponseEntity.ok(Map.of("username", username, "message", "Profile endpoint"));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error fetching profile: " + e.getMessage());
        }
    }
    
    @PostMapping("/profile")
    public ResponseEntity<?> updateProfile(@RequestBody Map<String, Object> profileData) {
        try {
            // Implement profile update logic
            return ResponseEntity.ok("Profile updated successfully");
        } catch (Exception e) {
            return ResponseEntity.badRequest().body("Error updating profile: " + e.getMessage());
        }
    }
    
    // Helper method to get tiffin ID from username
    private Long getTiffinIdFromUsername(String username) {
        try {
            Optional<Tiffin> tiffin = tiffinRepository.findByUsername(username);
            if (tiffin.isPresent()) {
                return tiffin.get().getId();
            }
            throw new RuntimeException("Tiffin not found for username: " + username);
        } catch (Exception e) {
            throw new RuntimeException("Error getting tiffin ID: " + e.getMessage());
        }
    }
} 

File: PGVaaleDotNetBackend/Controllers/TiffinController.cs
using Microsoft.AspNetCore.Mvc;
using PGVaaleDotNetBackend.DTOs;
using PGVaaleDotNetBackend.Entities;
using PGVaaleDotNetBackend.Repositories;
using PGVaaleDotNetBackend.Services;

namespace PGVaaleDotNetBackend.Controllers
{
    [ApiController]
    [Route("api/tiffin")]
    public class TiffinController : ControllerBase
    {
        private readonly TiffinService _tiffinService;
        private readonly ITiffinRepository _tiffinRepository;
        private readonly IFeedback_TiffinRepository _feedbackTiffinRepository;
        private readonly IUserTiffinRepository _userTiffinRepository;
        private readonly IMenuRepository _menuRepository;

        public TiffinController(
            TiffinService tiffinService,
            ITiffinRepository tiffinRepository,
            IFeedback_TiffinRepository feedbackTiffinRepository,
            IUserTiffinRepository userTiffinRepository,
            IMenuRepository menuRepository)
        {
            _tiffinService = tiffinService;
            _tiffinRepository = tiffinRepository;
            _feedbackTiffinRepository = feedbackTiffinRepository;
            _userTiffinRepository = userTiffinRepository;
            _menuRepository = menuRepository;
        }

        // Dashboard
        [HttpGet("dashboard")]
        public async Task<IActionResult> GetDashboard()
        {
            try
            {
                // TODO: Get tiffin ID from authentication context
                // For now, we'll use a placeholder - you'll need to implement proper authentication
                var tiffinId = await GetTiffinIdFromUsernameAsync("placeholder_username");
                var dashboard = await _tiffinService.GetTiffinDashboardAsync(tiffinId);
                return Ok(dashboard);
            }
            catch (Exception e)
            {
                return BadRequest($"Error fetching dashboard: {e.Message}");
            }
        }

        // Get all tiffins (for frontend)
        [HttpGet("all")]
        public async Task<IActionResult> GetAllTiffins()
        {
            try
            {
                var tiffins = await _tiffinService.GetAllTiffinsAsync();
                // Only return approved tiffins for public access
                var approvedTiffins = tiffins.Where(t => t.Approved).ToList();
                return Ok(approvedTiffins);
            }
            catch (Exception e)
            {
                return BadRequest($"Error fetching tiffins: {e.Message}");
            }
        }

        // Menu Management
        [HttpPost("menu")]
        public async Task<IActionResult> CreateMenu([FromBody] MenuDTO menuDTO)
        {
            try
            {
                // TODO: Get tiffin ID from authentication context
                var tiffinId = await GetTiffinIdFromUsernameAsync("placeholder_username");
                menuDTO.TiffinId = tiffinId;
                var createdMenu = await _tiffinService.CreateMenuAsync(menuDTO);
                return Ok(createdMenu);
            }
            catch (Exception e)
            {
                return BadRequest($"Error creating menu: {e.Message}");
            }
        }

        [HttpPut("menu/{menuId}")]
        public async Task<IActionResult> UpdateMenu(long menuId, [FromBody] MenuDTO menuDTO)
        {
            try
            {
                var updatedMenu = await _tiffinService.UpdateMenuAsync(menuId, menuDTO);
                return Ok(updatedMenu);
            }
            catch (Exception e)
            {
                return BadRequest($"Error updating menu: {e.Message}");
            }
        }

        [HttpDelete("menu/{menuId}")]
        public async Task<IActionResult> DeleteMenu(long menuId)
        {
            try
            {
                await _tiffinService.DeleteMenuAsync(menuId);
                return Ok("Menu deleted successfully");
            }
            catch (Exception e)
            {
                return BadRequest($"Error deleting menu: {e.Message}");
            }
        }

        [HttpGet("menu")]
        public async Task<IActionResult> GetWeeklyMenu()
        {
            try
            {
                // TODO: Get tiffin ID from authentication context
                var tiffinId = await GetTiffinIdFromUsernameAsync("placeholder_username");
                var menus = await _tiffinService.GetWeeklyMenuAsync(tiffinId);
                return Ok(menus);
            }
            catch (Exception e)
            {
                return BadRequest($"Error fetching menu: {e.Message}");
            }
        }

        [HttpGet("menu/{dayOfWeek}")]
        public async Task<IActionResult> GetMenuByDay(string dayOfWeek)
        {
            try
            {
                // TODO: Get tiffin ID from authentication context
                var tiffinId = await GetTiffinIdFromUsernameAsync("placeholder_username");
                var menu = await _tiffinService.GetMenuByDayAsync(tiffinId, dayOfWeek);
                if (menu != null)
                {
                    return Ok(menu);
                }
                else
                {
                    return NotFound();
                }
            }
            catch (Exception e)
            {
                return BadRequest($"Error fetching menu: {e.Message}");
            }
        }

        // Request Management
        [HttpGet("requests")]
        public async Task<IActionResult> GetRequests([FromQuery] string? status)
        {
            try
            {
                // TODO: Get tiffin ID from authentication context
                var tiffinId = await GetTiffinIdFromUsernameAsync("placeholder_username");
                UserTiffin.RequestStatus? requestStatus = null;
                if (!string.IsNullOrEmpty(status))
                {
                    if (Enum.TryParse<UserTiffin.RequestStatus>(status.ToUpper(), out var parsedStatus))
                    {
                        requestStatus = parsedStatus;
                    }
                }
                var requests = await _tiffinService.GetTiffinRequestsAsync(tiffinId, requestStatus);
                return Ok(requests);
            }
            catch (Exception e)
            {
                return BadRequest($"Error fetching requests: {e.Message}");
            }
        }

        [HttpPost("requests/{requestId}/status")]
        public async Task<IActionResult> UpdateRequestStatus(long requestId, [FromBody] Dictionary<string, string> request)
        {
            try
            {
                var status = request["status"];
                if (Enum.TryParse<UserTiffin.RequestStatus>(status.ToUpper(), out var requestStatus))
                {
                    var updatedRequest = await _tiffinService.UpdateRequestStatusAsync(requestId, requestStatus);
                    return Ok(updatedRequest);
                }
                else
                {
                    return BadRequest("Invalid status value");
                }
            }
            catch (Exception e)
            {
                return BadRequest($"Error updating request status: {e.Message}");
            }
        }

        // Profile
        [HttpGet("profile")]
        public async Task<IActionResult> GetProfile()
        {
            try
            {
                // TODO: Get username from authentication context
                var username = "placeholder_username";
                return Ok(new { username = username, message = "Profile endpoint" });
            }
            catch (Exception e)
            {
                return BadRequest($"Error fetching profile: {e.Message}");
            }
        }

        [HttpPost("profile")]
        public async Task<IActionResult> UpdateProfile([FromBody] Dictionary<string, object> profileData)
        {
            try
            {
                // TODO: Implement profile update logic
                return Ok("Profile updated successfully");
            }
            catch (Exception e)
            {
                return BadRequest($"Error updating profile: {e.Message}");
            }
        }

        // Delete tiffin account
        [HttpDelete("profile")]
        public async Task<IActionResult> DeleteAccount()
        {
            try
            {
                // TODO: Get username from authentication context
                var username = "placeholder_username";
                var tiffinOptional = await _tiffinRepository.FindByUsernameAsync(username);
                if (tiffinOptional == null)
                {
                    return NotFound();
                }

                var tiffin = tiffinOptional;
                var tiffinId = tiffin.Id;

                // Delete all feedback for this tiffin
                var tiffinFeedback = await _feedbackTiffinRepository.GetByTiffinIdAsync(tiffinId);
                foreach (var feedback in tiffinFeedback)
                {
                    _feedbackTiffinRepository.Delete(feedback.Id);
                }

                // Delete all user-tiffin relationships for this tiffin
                var userTiffinRelations = await _userTiffinRepository.FindByTiffinIdAsync(tiffinId);
                foreach (var relation in userTiffinRelations)
                {
                    await _userTiffinRepository.DeleteAsync(relation.Id);
                }

                // Delete all menus for this tiffin
                var tiffinMenus = await _menuRepository.FindByTiffinIdAndIsActiveTrueAsync(tiffinId);
                foreach (var menu in tiffinMenus)
                {
                    await _menuRepository.DeleteAsync(menu.Id);
                }

                // Delete the tiffin account
                await _tiffinRepository.DeleteAsync(tiffinId);

                return Ok("Account deleted successfully");
            }
            catch (Exception e)
            {
                return StatusCode(500, $"Error deleting account: {e.Message}");
            }
        }

        // Helper method to get tiffin ID from username
        private async Task<long> GetTiffinIdFromUsernameAsync(string username)
        {
            try
            {
                var tiffin = await _tiffinRepository.FindByUsernameAsync(username);
                if (tiffin != null)
                {
                    return tiffin.Id;
                }
                throw new InvalidOperationException($"Tiffin not found for username: {username}");
            }
            catch (Exception e)
            {
                throw new InvalidOperationException($"Error getting tiffin ID: {e.Message}");
            }
        }
    }
}
